x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java

```diff
@@ -249,11 +299,30 @@ private void onGapSuccess(final long start, final long end) {
             assert invariant();
         }
 
-        completionListener.onResponse(null);
+        completionListener.onResponse(end);
+    }
+
+    private void onGapProgress(long start, long end, long value) {
+        final ProgressListenableActionFuture completionListener;
+
+        synchronized (mutex) {
+            assert invariant();
+
+            final Range range = new Range(start, end, null);
+            final SortedSet<Range> existingRanges = ranges.tailSet(range);
```

Comment:
Seems strange to have to look up the range corresponding with the gap here, maybe `Gap` should keep hold of the corresponding `Range` so it can call the listener directly.

Reply:
The `onSuccess` and `onFailure` methods would also benefit from keeping a reference to the range; the look up should not be necessary there too as a pending `Range` should not be completed or failed outside of the corresponding `Gap`. I'm tempted to address this in a follow up PR.

Reply:
Sure, a followup is fine. The difference with `onSuccess` and `onFailure` is that they also adjust `ranges`, but they do indeed start with the same kind of lookup as we do here.

Reply:
I opened https://github.com/elastic/elasticsearch/pull/58587 for this.